### PR TITLE
Redirect System out and err to log4j (#840)

### PR DESCRIPTION
**Pull request by KyleFung**
Created a class under utils that will tie out and err to log4j. This
class does this in the main functions of AzkabanWebServer and
AzkabanExecServer, so that all out going messages will be caught by
log4j.

### CODE REVIEW

**Comment by kunkun-tang** [comment]
I'm confused that Ihow do you define "out going messages" ?

**Comment by KyleFung** [comment]
I meant any messages that the server tries to relay to the outside world will be caught by log4j. This includes exceptions and stack traces.

File: azkaban-common/src/main/java/azkaban/utils/StdOutErrRedirect.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  private static final Logger logger = Logger.getLogger(StdOutErrRedirect.class);
+
+  public static void redirectOutAndErrToLog() {
+    System.setOut(createStream(System.out, Level.INFO));
+    System.setErr(createStream(System.err, Level.ERROR));
+  }
+
+  private static PrintStream createStream(PrintStream stream, Level level) {
+    return new PrintStream(stream) {
+      public void println(final String string) {
```
Comment (suvodeep-pyne) [inline azkaban-common/src/main/java/azkaban/utils/StdOutErrRedirect.java]: Curious: Will this work if I print an integer ?

Best practices: if you are overriding/implementing a function, use annotations to indicate them. Like in this case using @Override.

File: azkaban-common/src/main/java/azkaban/utils/StdOutErrRedirect.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import java.io.PrintStream;
+
+import org.apache.log4j.Logger;
+import org.apache.log4j.Level;
+
+public class StdOutErrRedirect {
+  private static final Logger logger = Logger.getLogger(StdOutErrRedirect.class);
+
+  public static void redirectOutAndErrToLog() {
+    System.setOut(createStream(System.out, Level.INFO));
```
Comment (suvodeep-pyne) [inline azkaban-common/src/main/java/azkaban/utils/StdOutErrRedirect.java]: In the solo server case, there will be 2 PrintStream objects created. What do you think about maintaining a single instance of each in the class itself?

Do you see any advantages of creating a fresh stream every time a redirect is called ?

File: azkaban-solo-server/src/main/resources/log4j.properties [hunk 3]
```diff
@@ -1,7 +1,5 @@
 log4j.rootLogger=INFO, Console
-log4j.logger.azkaban.webapp=INFO, server
-log4j.logger.azkaban.execapp=INFO, server
-log4j.logger.azkaban.soloserver=INFO, server
+log4j.logger.azkaban=INFO, server
```
Comment (suvodeep-pyne) [inline azkaban-solo-server/src/main/resources/log4j.properties]: Thanks for tidying this up

File: azkaban-common/src/main/java/azkaban/utils/StdOutErrRedirect.java (hunk 1 above)
Comment (KyleFung) [inline azkaban-common/src/main/java/azkaban/utils/StdOutErrRedirect.java]: It seems you're right about the integer.

File: azkaban-common/src/main/java/azkaban/utils/StdOutErrRedirect.java (hunk 2 above)
Comment (KyleFung) [inline azkaban-common/src/main/java/azkaban/utils/StdOutErrRedirect.java]: Latest version creates static instances of them.

File: azkaban-common/src/main/java/azkaban/utils/StdOutErrRedirect.java (hunk 1 above)
Comment (KyleFung) [inline azkaban-common/src/main/java/azkaban/utils/StdOutErrRedirect.java]: Fixed by overriding a bunch of print/println functions. Unfortunately this was the only way, apart from overriding write(), but that option seemed very error prone and not as simple as overriding print/println.

File: azkaban-common/src/main/java/azkaban/utils/StdOutErrRedirect.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+/**
+ * A class to encapsulate the redirection of stdout and stderr to log4j
+ * This allows us to catch messages written to the console (although we should
+ * not be using System.out to write out).
+ */
+
+public class StdOutErrRedirect {
+  private static final Logger logger = Logger.getLogger(StdOutErrRedirect.class);
+  private static PrintStream infoStream = createStream(System.out, Level.INFO);
```
Comment (suvodeep-pyne) [inline azkaban-common/src/main/java/azkaban/utils/StdOutErrRedirect.java]: preferably final ?

File: azkaban-common/src/main/java/azkaban/utils/StdOutErrRedirect.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    System.setOut(infoStream);
+    System.setErr(errorStream);
+  }
+
+  private static PrintStream createStream(PrintStream stream, Level level) {
+    return new LogStream(stream, level);
+  };
+
+  private static class LogStream extends PrintStream {
+    private Level level;
```
Comment (suvodeep-pyne) [inline azkaban-common/src/main/java/azkaban/utils/StdOutErrRedirect.java]: nit: Let's keep variables final whenever they can be

File: azkaban-common/src/main/java/azkaban/utils/StdOutErrRedirect.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+      this.level = level;
+    }
+    // String - all other overridden methods will call this
+    @Override
+    public void println(final String string) {
+      print(string);
+    }
+    @Override
+    public void print(final String string) {
+      logger.log(level, string);
```
Comment (suvodeep-pyne) [inline azkaban-common/src/main/java/azkaban/utils/StdOutErrRedirect.java]: Can we have a separate differently named private method which does this, similar to the OutputStream in java. That makes the code much more readable than calling an already heavily overloaded and overridden method.

File: azkaban-common/src/main/java/azkaban/utils/StdOutErrRedirect.java [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+  };
+
+  private static class LogStream extends PrintStream {
+    private Level level;
+
+    public LogStream(OutputStream out, Level level) {
+      super(out);
+      this.level = level;
+    }
+    // String - all other overridden methods will call this
```
Comment (suvodeep-pyne) [inline azkaban-common/src/main/java/azkaban/utils/StdOutErrRedirect.java]: nit: formatting: single line gap between methods

**Pull request merged**